pkg/service/backup/restore.go

```diff
@@ -72,8 +72,97 @@ func (s *Service) GetRestoreTarget(ctx context.Context, clusterID uuid.UUID, pro
 	return t, nil
 }
 
+// Restore executes restore on a given target.
 func (s *Service) Restore(ctx context.Context, clusterID, taskID, runID uuid.UUID, target RestoreTarget) error {
-	panic("TODO - implement")
+	s.logger.Info(ctx, "Restore",
+		"cluster_id", clusterID,
+		"task_id", taskID,
+		"run_id", runID,
+		"target", target,
+	)
+
+	run := &RestoreRun{
+		ClusterID:   clusterID,
+		TaskID:      taskID,
+		ID:          runID,
+		SnapshotTag: target.SnapshotTag,
+		Stage:       StageRestoreInit,
+	}
+
+	// Get cluster name
+	clusterName, err := s.clusterName(ctx, clusterID)
+	if err != nil {
+		return errors.Wrap(err, "invalid cluster")
+	}
+	// Get the cluster client
+	client, err := s.scyllaClient(ctx, clusterID)
+	if err != nil {
+		return errors.Wrap(err, "get client proxy")
+	}
+	// Get cluster session
+	clusterSession, err := s.clusterSession(ctx, clusterID)
+	if err != nil {
+		return errors.Wrap(err, "get CQL cluster session")
+	}
+	defer clusterSession.Close()
+
+	w := &restoreWorker{
+		worker: worker{
+			ClusterID:   clusterID,
+			ClusterName: clusterName,
+			TaskID:      taskID,
+			RunID:       runID,
+			Client:      client,
+			Config:      s.config,
+			Metrics:     s.metrics,
+			Logger:      s.logger.Named("restore"),
+		},
+		managerSession:          s.session,
+		clusterSession:          clusterSession,
+		forEachRestoredManifest: s.forEachRestoredManifest(clusterID, target),
+	}
+
+	if target.Continue {
+		if err := w.decorateWithPrevRun(ctx, run); err != nil {
+			return err
+		}
+
+		w.InsertRun(ctx, run)
+		// Update run with previous progress.
+		if run.PrevID != uuid.Nil {
+			w.clonePrevProgress(ctx, run)
+		}
+
+		w.Logger.Info(ctx, "Run after decoration",
+			"run", *run,
+		)
+	} else {
+		w.InsertRun(ctx, run)
+	}
+
+	if !target.Continue || run.PrevID == uuid.Nil {
+		w.resumed = true
+	}
+
+	// As manifests are immutable, units can be initialized only once per task
+	if run.Units == nil {
+		run.Units, err = w.newUnits(ctx, target)
+		if err != nil {
+			return errors.Wrap(err, "initialize units")
+		}
+	}
+
+	run.Stage = StageRestoreData
+	w.InsertRun(ctx, run)
+
+	if err := w.restoreData(ctx, run, target); err != nil {
+		return errors.Wrapf(err, "restore data")
```

Comment:
In what state the ScyllaDB is left if an error during restore appears ?
Is it something that can be reversed ?
Just wondering whether there should be some more descriptive information on stdout when an error appeard. I mean what to do to fix the state.

Reply:
In case of error, restore should behave in the same way as in case of context cancellation (so a safe state that has some already restored data and can be continued - at least that's the goal). Problem occurs if error is returned from resetting gc_grace_seconds and compaction on the table, because then table will stay in this disabled state. For this scenario I can add more descriptive error message which will include original values of gc_grace_seconds and compaction and ask user to reset them manually.

But that's from scylla-manager perspective.

I don't know how rclone handles errors (does it actually cleans already downloaded data or does it leave it there?).
The same goes for error during load and stream procedure (does it clean upload directory, does it clean already uploaded data to Scylla?)

What do you think about that?

Reply:
> Problem occurs if error is returned from resetting gc_grace_seconds and compaction on the table, because then table will stay in this disabled state. For this scenario I can add more descriptive error message which will include original values of gc_grace_seconds and compaction and ask user to reset them manually.

Yes, please.

> I don't know how rclone handles errors (does it actually cleans already downloaded data or does it leave it there?).
The same goes for error during load and stream procedure (does it clean upload directory, does it clean already uploaded data to Scylla?)

To be asked on the planning.